### PR TITLE
Backport 3.6: ssl_client1 and ssl_server demo scripts

### LINKED ISSUE #9072: ssl_client1 fails on TLS 1.3
### Summary
If I attempt to make a request to `https://api.sunrisesunset.io/json?lat=47.333&lng=13.333` using Postman, or a browser, it works perfectly.

However, when making an HTTP request using the mbedTLS sample client [ssl_client1.c](https://github.com/Mbed-TLS/mbedtls/blob/development/programs/ssl/ssl_client1.c) with these parameters:

```
#define SERVER_PORT "443"
#define SERVER_NAME "api.sunrisesunset.io"
#define GET_REQUEST "GET /json?lat=47.333&lng=13.333 HTTP/1.1\r\nHost: api.sunrisesunset.io\r\n\r\n"
```

I got the following error:

```
ssl_client.c:0042: client hello, adding server name extension: api.sunrisesunset.io
ssl_tls13_client.c:0057: client hello, adding supported versions extension
ssl_tls13_client.c:0080: supported version: [3:4]
ssl_tls13_client.c:0086: supported version: [3:3]
ssl_tls13_client.c:0572: no cookie to send; skip extension
ssl_tls13_client.c:0285: client hello: adding key share extension
ssl_tls13_generic.c:1651: Perform PSA-based ECDH/FFDH computation.
ssl_tls13_generic.c:1689: psa_generate_key() returned -27648 (-0x6c00)
ssl_client.c:1012: <= write client hello
ssl_tls.c:4617: <= handshake
 failed
  ! mbedtls_ssl_handshake returned -0x6c00

Last error was: -27648 - SSL - Internal error (eg, unexpected failure in lower-level module)
```

What is interesting is that when I force the mbedTLS to use TLS 1.2, instead TLS 1.3 by adding the following:

`mbedtls_ssl_conf_max_version(&conf, MBEDTLS_SSL_MAJOR_VERSION_3, MBEDTLS_SSL_MINOR_VERSION_3);  // Force TLS 1.2`

The request works perfectly! So, it indicates to me that there might be a problem with the TLS 1.3, somehow the server is not properly working with mbedTLS TLS 1.3.

### System information

Mbed TLS version (number or commit id): `2ca6c285a0dd3f33982dd57299012dacab1ff206`
Operating system and version: `macOS 13.2.1 (22D68)`
Configuration (if not default, please attach `mbedtls_config.h`): `default` 

### Expected behavior
The request should return a successful result.

### Actual behavior
The request is failing during the handshake as previously mentioned:
```
ssl_tls.c:4617: <= handshake
 failed
  ! mbedtls_ssl_handshake returned -0x6c00

Last error was: -27648 - SSL - Internal error (eg, unexpected failure in lower-level module)
```

### Steps to reproduce
Just change these configs on [ssl_client1.c](https://github.com/Mbed-TLS/mbedtls/blob/development/programs/ssl/ssl_client1.c) sample:

```
#define SERVER_PORT "443"
#define SERVER_NAME "api.sunrisesunset.io"
#define GET_REQUEST "GET /json?lat=47.333&lng=13.333 HTTP/1.1\r\nHost: api.sunrisesunset.io\r\n\r\n"
```

Then, rebuild and run that.

### Additional information
- It's working normally when calling from a browser, so it's not a server issue.
- It seems to be solved by forcing the usage of TLS 1.2, as previously mentioned:
`mbedtls_ssl_conf_max_version(&conf, MBEDTLS_SSL_MAJOR_VERSION_3, MBEDTLS_SSL_MINOR_VERSION_3);`

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
Add demo scripts for `ssl_client1`, `ssl_server`, `ssl_pthread_server` and `dtls_server`. This way we have CI coverage for:

* These sample programs.
* A no-frills, out-of-box TLS connection.

The TLS 1.3 demo reveals https://github.com/Mbed-TLS/mbedtls/issues/9072 and https://github.com/Mbed-TLS/mbedtls/issues/8749.

Fixes #9072 in the sense that when this pull request passes the CI, it means the issue is fixed. (The actual fixes are in previous PR: #9501 and #9507.)

The sample programs hard-code port 4433, so the demos can't run concurrently. This may be a problem for CI reliability (although in principle it shouldn't be since each test run runs alone in a Docker instance).

Omitted:

* `mini_client` — would require different interactions with `openssl s_server`.
* `dtls_client` — would require different interactions with `openssl s_server`.
* `ssl_fork_server` — would require fancier process handling.

Priority: high because it is a non-regression test for https://github.com/Mbed-TLS/mbedtls/issues/9072 and https://github.com/Mbed-TLS/mbedtls/issues/8749. Not very-high because I don't propose to rush this into 3.6.1. We can do manual testing for 3.6.1, and have the automated testing before 3.6.2.

Status: I expect a CI failure until https://github.com/Mbed-TLS/mbedtls/issues/8749 is fixed.

## PR checklist

- [x] **changelog** not required because: only tests
- [ ] **development PR** TODO
- [x] **framework PR** not required
- [x] **3.6 PR** here
- [ ] **2.28 PR** TODO
- **tests**  provided

### CODE REVIEW

File: programs/ssl/tls_client_demo_common.sh [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    ret=0
+    printf '+ %s\n' "$program"
+    "$program" || ret=$?
+    kill "$server_pid" 2>/dev/null || true # The server may exit first
+    # Check and display the presence of a few connection parameters
+    grep '^ *client_version=' "$server_log"
+    grep '^ *KeyExchangeAlgorithm=' "$server_log" || # TLS 1.2
+        grep -A1 '^ *extension_type=key_share' "$server_log" # TLS 1.3
+    grep '^ *cipher_suite ' "$server_log"
+    grep 'ApplicationData' "$server_log"
```
Comment (gilles-peskine-arm) [inline programs/ssl/tls_client_demo_common.sh:65]: I should give better feedback if one of these fails.

File: programs/ssl/tls_client_demo_common.sh [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+. "${0%/*}/../demo_common.sh"
+
+msg <<'EOF'
+This script demonstrates the interoperability between a very simple
+$protocol client using Mbed TLS and an OpenSSL server.
+
+EOF
+
+if ! openssl version >/dev/null; then
```
Comment (gilles-peskine-arm) [inline programs/ssl/tls_client_demo_common.sh:19]: So far, I've only tested with OpenSSL 3.0.2 on Ubuntu 22.04.

**Comment by mpg** [comment]
@gilles-peskine-arm Is now a good time for me to start reviewing or should I wait for further work (like rebasing to get a green CI) first?

**Comment by gilles-peskine-arm** [comment]
I'm not very happy with the stability of the shell scripts. I fear that they might run into a race condition, or fail because they insist on port 4433. That being said, I started to rewrite this in Python, and I find it actually harder: it would take more time than I care to throw at this task. So I've changed my mind on the design: I want to try this as is, and if it turns out to be unstable on the CI, we'll either fix it in a hurry or temporarily disable it.

I'm going to rebase on top of the 3.6.1 release, to give it a chance to pass the CI.

**Comment by gilles-peskine-arm** [comment]
Well, that's not good. The SSL demos are failing almost systematically on the CI ([`tls13_server_demo.sh` passed once](https://mbedtls.trustedfirmware.org/job/mbed-tls-pr-head/job/PR-9505-head/3/execution/node/1301/log/)). They work on my machine.

I can't see an obvious cause in the logs. Sometimes the server can't bind port 4433, which suggests that the port is in use and would require the ability to change the port. But sometimes the connection seems to happen correctly and yet the demo is reported as failed.

Reusing `compat.sh` is now seeming more attractive.

**Comment by mpg** [comment]
> Reusing `compat.sh` is now seeming more attractive.

I had a superficial look yesterday and was wondering why not make this part of `ssl-opt.sh`. Seeing things like `sleep 1` and `sleep 2` immediately made me nervous and think that we solved those problems already.

(I have no strong opinion between `compat.sh` and `ssl-opt.sh` especially as in the long run I think they should be merged, but indeed I think we should take advantage of existing tools.)

**Comment by gilles-peskine-arm** [comment]
Superseded by https://github.com/Mbed-TLS/mbedtls/pull/9541. Manuel's suggestion of `ssl-opt.sh` is clearly the way to go.

**Pull request closed without merging**